[quant][graphmode] PrepareQuant step

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24428 [quant][graphmode] Integration tests for initial quantization graph mode
* #24427 [quant][graphmode] quant_fusion jit pass
* #24426 [quant][graphmode] insert_quant_dequant jit pass
* **#24425 [quant][graphmode] PreapreQuant step**

Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16933120](https://our.internmc.facebook.com/intern/diff/D16933120)

jerryzh168 committed:
[quant][graphmode] PreapreQuant step

Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] PreapreQuant step"


Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] PreapreQuant step"


Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -145,6 +146,32 @@ Node* createIntReprNode(Value* v, Graph* g) {
+// Create observer.forward Node and insert a call to observer forward function
+Node* insertObserverForwardCall(Value* v, Graph* g, script::Module module, script::Module observer_module) {
+  std::string observer_name = "observer_for_" + v->debugName();
+  module.register_module(observer_name, observer_module.clone());
```
We register a `observer_module.clone()` here, but then use the original `observer_module` below. We should use what `observer_module.clone()` returned as those two are different instances.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -145,6 +146,32 @@ Node* createIntReprNode(Value* v, Graph* g) {
+Node* insertObserverForwardCall(Value* v, Graph* g, script::Module module, script::Module observer_module) {
+  std::string observer_name = "observer_for_" + v->debugName();
+  module.register_module(observer_name, observer_module.clone());
+  WithInsertPoint ins(v->node());
```
You don't need to use `WithInsertPoint` if you use `insertAfter`. `WithInsertPoint` only affects `insert` function, `insertBefore` and `insertAfter` explicitly specify where the node should be inserted.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -145,6 +146,32 @@ Node* createIntReprNode(Value* v, Graph* g) {
+  observer_instance->insertAfter(v->node());
+
+  // Create forward method call
+  Node* call = observer_instance->owningGraph()->create(c10::prim::CallMethod);
```
Nit: you could use `g` instead of `observer_instance->owningGraph()`.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.h:
```diff
@@ -102,5 +102,12 @@ TORCH_API void InsertQuantDequantNodesForParam(
     const Fn& getQParamFunc,
     at::ScalarType t);
+
+TORCH_API script::Module PrepareQuant(
```
Please add documentation for new API.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -572,5 +599,107 @@ template TORCH_API void InsertQuantDequantNodesForParam(
+  return observerNode;
+}
+
+TORCH_API script::Module PrepareQuant(
```
I think the name of this function should be more specific - "prepare" can mean anything. I realize that it comes from how we describe it to users as they go through quantization process, but I think here in the code it doesn't have to use the same one.

jerryzh168 replied:
ok, I'll change to insertobserver

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -572,5 +599,107 @@ template TORCH_API void InsertQuantDequantNodesForParam(
+  TORCH_CHECK(graph != nullptr);
+  // num_activation_inputs is the number of activations or external data
+  // excluding the parameters
+  TORCH_CHECK(num_activation_inputs <= graph->inputs().size());
```
This check is useless since `method.num_inputs()` and `method.graph().num_inputs()` are the same. That said, the first argument of a method is `%self`, which probably doesn't need to be observed.

jerryzh168 replied:
> That said, the first argument of a method is %self, which probably doesn't need to be observed.
What do you mean?

ZolotukhinM replied:
The first argument of a method is always `%self` (the parent module). It is never a tensor so we never observe it. The check below handles it, but that check will always result in "do not observe" for the first argument, so we can just start from the second argument. I don't know if it's worth changing though.

jerryzh168 replied:
oh, sure, I can change that

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -572,5 +599,107 @@ template TORCH_API void InsertQuantDequantNodesForParam(
+  // prim::Param nodes do not belong to the graph. Hence the Insert
+  // point is the beginning of graph node. This also safe guards against
+  // observing a potentially mutated value due to some in-place operation
+  Node* insert_node = *graph->nodes().begin();
```
This is not used.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -572,5 +599,107 @@ template TORCH_API void InsertQuantDequantNodesForParam(
     at::ScalarType t);
 
+
+static Node* prepQuantAddObserverFor(
```
This doesn't seem to be used anywhere.

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1094,6 +1094,40 @@ def getQParamFunc(input_scale, weight_scale):
+
+        m = torch.jit.script(M())
+        observer = torch.jit.script(Observer())
+        torch._C._jit_pass_constant_propagation(m.graph)
```
I'd rather explicitly specify in which graph we're running const-prop (`m._c._get_method("forward").graph`). Also, in this test we probably don't even need to run it.

jerryzh168 replied:
we do assume the graph we modify is after constant prop, it might work without constprop in this test case though.

jerryzh168 replied:
actually, this is required. otherwise we will have many observers for redundant nodes.

ZolotukhinM reviewed (approved):
Mostly looks good, please find some comments inline!

jerryzh168 committed:
Update on "[quant][graphmode] PreapreQuant step"


Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] PreapreQuant step"


Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16933120](https://our.internmc.facebook.com/intern/diff/D16933120)

jerryzh168 committed:
Update on "[quant][graphmode] PreapreQuant step"


Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16933120](https://our.internmc.facebook.com/intern/diff/D16933120)

nishantpdce commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -572,5 +600,80 @@ template TORCH_API void InsertQuantDequantNodesForParam(
     at::ScalarType t);
 
+
+TORCH_API script::Module InsertObservers(
```
Do you eventually plan to remove the older InsertObserver method because i believe you would mostly be operating at Module level. I see there is some code duplication

jerryzh168 replied:
yes I will do that in separate PRs

jerryzh168 committed:
Update on "[quant][graphmode] PreapreQuant step"


Summary:
- _jit_pass_prepare_quant: clone the observer module in argument and insert that to the
module we want to quantize, insert observer calls for the Tensor values we want to observe

Test Plan:
python test/test_jit.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D16933120](https://our.internmc.facebook.com/intern/diff/D16933120)

raghuramank100 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -572,5 +601,80 @@ template TORCH_API void InsertQuantDequantNodesForParam(
+  // point is the beginning of graph node. This also safe guards against
+  // observing a potentially mutated value due to some in-place operation
+  for (size_t idx = 1; idx < method.num_inputs(); ++idx) {
+    auto& v = graph->inputs()[idx];
```
Are we always inserting observers at inputs? This should be optional, with default = True.

jerryzh168 replied:
we can consider corner cases later, this is a primitive version that makes many assumptions
